kloppy/infra/serializers/event/wyscout/deserializer_v3.py

```diff
@@ -672,6 +672,75 @@ def _parse_period_id(raw_period: str) -> int:
     return period_id
 
 
+def insert_substitution_events(
+    deserializer, events, raw_events, players, teams, periods, transformer
+):
+    # Step 1: Create substitution events
+    substitution_events = []
+    for team_id, periods_subs in raw_events["substitutions"].items():
+        for raw_period, sub_info in periods_subs.items():
+            for raw_seconds, players_info in sub_info.items():
+                subs_out = players_info["out"]
+                subs_in = players_info["in"]
+                for sub_out, sub_in in zip(subs_out, subs_in):
+                    sub_out_player = players[team_id][str(sub_out["playerId"])]
+                    sub_in_player = players[team_id][str(sub_in["playerId"])]
+
+                    # Build the substitution event
+                    sub_event = deserializer.event_factory.build_substitution(
+                        event_id=f"substitution-{sub_out['playerId']}-{sub_in['playerId']}",
+                        ball_owning_team=None,
+                        ball_state=None,
+                        coordinates=Point(x=0, y=0),
+                        player=sub_out_player,
+                        replacement_player=sub_in_player,
+                        team=teams[team_id],
+                        period=periods[int(raw_period[0]) - 1],
+                        timestamp=timedelta(seconds=int(raw_seconds)),
+                        result=None,
+                        raw_event=None,
+                        qualifiers=None,
+                    )
+
+                    if sub_event and deserializer.should_include_event(
+                        sub_event
+                    ):
+                        substitution_events.append(
+                            transformer.transform_event(sub_event)
+                        )
+
+    # Step 2: Sort substitution events globally by period and timestamp
+    substitution_events.sort(key=lambda e: (e.period.id, e.timestamp))
+
+    # Step 3: Merge events and substitutions in ascending order
+    merged_events = []
+    sub_index = 0
+    total_subs = len(substitution_events)
+
+    for event in events:
+        # Insert all substitution events that occur before or at the current event's timestamp
+        while sub_index < total_subs:
+            sub_event = substitution_events[sub_index]
+            if sub_event.period.id < event.period.id or (
+                sub_event.period.id == event.period.id
+                and sub_event.timestamp <= event.timestamp
+            ):
+                merged_events.append(sub_event)
+                sub_index += 1
+            else:
+                break
+
+        # Add the current event to the merged list
+        merged_events.append(event)
+
+    # Step 4: Add any remaining substitution events
+    while sub_index < total_subs:
+        merged_events.append(substitution_events[sub_index])
+        sub_index += 1
+
+    return merged_events
```

Comment:
Thanks for the help. I've refactored the code to use your more simple design.

In the test file that we use, I did not find any regular events that occur in the same second as the substitutions. But I'm indeed not sure, whether this will be generally true.

Do you think we should already start with merging this implementation of adding substitution events for the Wyscout V3 parser, and later in a separate issue/PR work on better handling the edge case of when a substitution event and a regular event happen on the same second?

Reply:
@probberechts do you think we can already merge this PR? And in a separate issue/PR work on better handling the edge case of when a substitution event and a regular event happen on the same second (if this even ever occurs)?